File: src/main/java/eu/execom/fabut/graph/NodesList.java
package eu.execom.fabut.graph;

import java.util.LinkedList;
import java.util.List;

import eu.execom.fabut.enums.NodeCheckType;
import eu.execom.fabut.pair.AssertPair;

/**
 * Implementing class for {@link IsomorphicGraph} using {@link LinkedList} as container.
 * 
 * @author Dusko Vesin
 * @author Nikola Olah
 * @author Bojan Babic
 * @author Nikola Trkulja
 */
public class NodesList implements IsomorphicGraph {
    private final List<IsomorphicNodePair> isomorphicNodes;

    /**
     * Default constructor.
     */
    public NodesList() {
        isomorphicNodes = new LinkedList<IsomorphicNodePair>();
    }

    @Override
    public boolean containsPair(final Object expected, final Object actual) {
        return isomorphicNodes.contains(new IsomorphicNodePair(expected, actual));
    }

    @Override
    public void addPair(final Object expected, final Object actual) {
        isomorphicNodes.add(new IsomorphicNodePair(expected, actual));
    }

    // FIXME this method is only used in tests, we can remove it
    public void addPair(final AssertPair assertPair) {
        addPair(assertPair.getExpected(), assertPair.getActual());
    }

    @SuppressWarnings("unchecked")
    @Override
    public Object getExpected(final Object actual) {
        for (final IsomorphicNodePair isomorphicNode : isomorphicNodes) {
            if (isomorphicNode.getActual() == actual) {
                return isomorphicNode.getExpected();
            }
        }
        return null;
    }

    @Override
    public boolean containsActual(final Object actual) {
        for (final IsomorphicNodePair isomorphicNode : isomorphicNodes) {
            if (isomorphicNode.getActual() == actual) {
                return true;
            }
        }
        return false;
    }

    @Override
    public boolean containsExpected(final Object expected) {
        for (final IsomorphicNodePair isomorphicNode : isomorphicNodes) {
            if (isomorphicNode.getExpected() == expected) {
                return true;
            }
        }
        return false;
    }

    @Override
    public NodeCheckType nodeCheck(final Object expected, final Object actual) {
        if (containsPair(expected, actual)) {
            return NodeCheckType.CONTAINS_PAIR;
        } else if (containsExpected(actual) || containsActual(expected)) {
            return NodeCheckType.SINGLE_NODE;
        }
        return NodeCheckType.NEW_PAIR;
    }

    public NodeCheckType nodeCheck(final AssertPair assertPair) {
        return nodeCheck(assertPair.getExpected(), assertPair.getActual());
    }

}

File: src/main/java/eu/execom/fabut/graph/NodesList.scala
package eu.execom.fabut.graph

import eu.execom.fabut.pair.AssertPair
import eu.execom.fabut.enums.NodeCheckType._

class NodesList extends IsomorphicGraph {

  var isomorphicNodes: List[IsomorphicNodePair] = List()

  override def addPair(expected: AnyRef, actual: AnyRef) {
    isomorphicNodes ::= IsomorphicNodePair(expected, actual)
  }

  override def containsPair(expected: AnyRef, actual: AnyRef) = {
    isomorphicNodes contains (IsomorphicNodePair(expected, actual))
  }

  override def containsExpected(expected: AnyRef) = {
    val pair = isomorphicNodes find (pair => pair.expected == expected)

    if (pair.isDefined) true else false
  }

  //  override def containsActual(actual: Any) = {
  //    val pair = isomorphicNodes find (pair => pair.actual == actual)
  //
  //    if (pair.isDefined) true else false
  //  }

  override def expected(actual: AnyRef): Option[AnyRef] = {
    val pair = isomorphicNodes find (pair => pair.actual.asInstanceOf[AnyRef] eq actual)

    if (pair.isDefined) Some(pair.get.expected.asInstanceOf[AnyRef]) else None
  }

  //  override def getActual(expected: Any): Option[Any] = {
  //    val pair = isomorphicNodes find (pair => pair.expected == expected)
  //
  //    if (pair.isDefined) Some(pair.get.actual) else None
  //  }

  override def nodeCheck(pair: AssertPair): NodeCheckType = {

    if (containsExpected(pair.expected.asInstanceOf[AnyRef])) {
      CONTAINS_PAIR
    } else {
      NEW_PAIR
    }

  }

}